### DISCUSSION TITLE
Zoom

### DISCUSSION THREAD

**Discussion by BoscoBecker** [graphics32/graphics32#328] [General] [answered]
Actually, it's a question

How could I create a magnifying glass effect on an image and create a zoom effect, like there is in e-commerce.

example below:

![image](https://github.com/user-attachments/assets/14a0b374-a3ae-49c6-9a7b-8c6883309470)

Site: https://www.magictoolbox.com/magiczoomplus/integration/

**Comment by andersmelander** [comment]
You can do that with a `TImage32`; You control the zoom with the `Scale` property and the position with the `OffsetHorz` and `OffsetVert` properties.

The hard part is figuring out what the values of `OffsetHorz` and `OffsetVert` must be based on the mouse position. I will create a small example once I figure it out myself :-)

**Reply by BoscoBecker** [reply]
Wow, perfect, I was waiting for this, I looked everywhere but what I found was the Windows magnifying glass effect (native).

Something like this: https://zarko-gajic.iz.hr/gzoom-delphi-implementation-of-the-missing-mode-in-windows-magnifier/

**Reply by andersmelander** [reply]
The zoom part is very easy (see the "ImgView_Layers" and "Pan & Zoom" examples). Panning the view is also easy.

The (very) challenging part is panning the view so the zoomed mouse position corresponds exactly to the unzoomed position, like in the [Internal Zoom example](https://www.magictoolbox.com/magiczoomplus/integration/#inner-zoom) - but I'm almost there.

**Comment by andersmelander** [comment]
I've gotten it to work for the case where the zoom increases the scale up **from** 1 but I still can't make it work for the case where the zoom increases the scale up **to** 1 (as is the case for the shoe).

https://github.com/user-attachments/assets/0018ee4a-10bc-4a5d-be46-e18dfd38d4ef

https://github.com/user-attachments/assets/9a31c041-b550-42db-b2f8-4be681a6234b

**Comment by BoscoBecker** [comment]
Apparently these web effects are generally easier than Delphi or other. But how cool, that would solve it, would you have a demo?

Thank you very much.

**Reply by andersmelander** [reply]
They are easy when you have a library (such as Magic Zoom Plus) that already implement them but somebody had to write that library in the first place.

I will commit the demo when I have solved the last part.

**Comment by BoscoBecker** [comment]
Thanks a lot .

**Comment by andersmelander** [comment, marked as answer]
Now it works for both small and large images.
![billede](https://github.com/user-attachments/assets/3c556243-ea8f-4eef-8895-92906e3f7122)


The demo is the [HoverZoom](https://github.com/graphics32/graphics32/tree/master/Examples/General/HoverZoom) example. Give it a spin.

**Comment by BoscoBecker** [comment]
Wow, wonderful, I'll test the example, I'll probably create a component, so about the example. As soon as I finish it I'll leave the link here and add the credits. Thank you very much for the support.
